concordium-consensus/src/Concordium/KonsensusV1/Consensus.hs

```diff
@@ -31,4 +26,67 @@ class MonadMulticast m where
 newtype BakerContext = BakerContext
     { _bakerIdentity :: BakerIdentity
     }
-makeClassy ''BakerContext
\ No newline at end of file
+
+makeClassy ''BakerContext
+
+-- |A Monad for timer related actions.
+class MonadTimeout m where
+    -- |Reset the timeout from the supplied 'Duration'.
+    resetTimer :: Duration -> m ()
+
+-- |Produce a block and multicast it onto the network.
+makeBlock :: MonadState (SkovData (MPV m)) m => m ()
+makeBlock = return ()
+
+-- |Make a block if the consensus runner is leader for the
+-- current round.
+-- TODO: call 'makeBlock' if we're leader for the current round.
+makeBlockIfLeader :: MonadState (SkovData (MPV m)) m => m ()
+makeBlockIfLeader = return ()
+
+-- |Advance to the provided 'Round'.
+--
+-- This function does the following:
+-- * Update the current 'RoundStatus'.
+-- * Persist the new 'RoundStatus'.
+-- * If the consensus runner is leader in the new
+--   round then make the new block.
+advanceRound ::
+    ( MonadReader r m,
+      HasBakerContext r,
+      MonadTimeout m,
+      LowLevel.MonadTreeStateStore m,
+      MonadState (SkovData (MPV m)) m
+    ) =>
+    -- |The 'Round' to progress to.
+    Round ->
+    -- |If we are advancing from a round that timed out
+    -- then this will be @Left 'TimeoutCertificate, 'QuorumCertificate')@
+    -- The 'TimeoutCertificate' is from the round we're
+    -- advancing from and the associated 'QuorumCertificate' verifies it.
+    --
+    -- Otherwise if we're progressing via a 'QuorumCertificate' then @Right QuorumCertificate@
+    -- should be the QC we're advancing round via.
+    Either (TimeoutCertificate, QuorumCertificate) QuorumCertificate ->
+    m ()
+advanceRound newRound timedOut = do
+    myBakerId <- bakerId <$> view bakerIdentity
+    currentRoundStatus <- use roundStatus
+    -- Reset the timeout timer if the consensus runner is part of the
+    -- finalization committee.
+    resetTimerIfFinalizer myBakerId (rsCurrentTimeout currentRoundStatus) (rsCurrentEpoch currentRoundStatus)
+    -- Advance and save the round.
+    setRoundStatus $! advanceRoundStatus newRound timedOut currentRoundStatus
+    -- Make a new block if the consensus runner is leader of
+    -- the 'Round' progressed to.
+    makeBlockIfLeader
+  where
+    -- Reset the timer if this consensus instance is member of the
+    -- finalization committee for the current 'Epoch'.
+    resetTimerIfFinalizer bakerId currentTimeout currentEpoch = do
+        gets (getBakersForLiveEpoch currentEpoch) >>= \case
+            Nothing -> return () -- No bakers or finalizers could be looked up for the current 'Epoch' so we do nothing.
+            Just bakersAndFinalizers -> do
+                when (isJust $ finalizerByBakerId (bakersAndFinalizers ^. bfFinalizers) bakerId) $
+                    -- The consensus runner is a finalizer for the current epoch then we reset the timer
+                    resetTimer currentTimeout
```

Comment:
A problem here this is that people leaving the finalization committee might not start a timer. But if their highestQC is still in the previous epoch, we would like these people to throw the timeout. So they have to start the timer for that.

The bluepaper is probably wrong in this regard. I did not consider edge cases of people joining and leaving the finalization committee when writing it up. It is complicated enough as it is, and I prefer leave those cases to you.

There is a similar issue with people joining the finalization committee. Being a finalizer is not enough to be able to throw a timeout when the timer expires. We also have to have a highestQC in an epoch where we are a finalizer. One option is to take care of this is when throwing the timeout (`uponTimeoutEvent` in the bluepaper). An alternative is to not even start a timer in that case. The first is probably the better option, because the highestQC can change between the moment were we start the timer and when it expires. But you might also want it to be consistent with the case of people leaving, and start timers based on the epoch of the highestQC and not of the currentepoch. Anyway, since I haven't seen the code for `uponTimeoutEvent` yet, I am just pointing this out here so that you can check what is being done.

Note that I wrote a post in slack about the case of people joining: https://concordium.slack.com/archives/CGFCLTN8Z/p1677852233604619

Reply:
Always resetting the timer now. I've put some documentation that emphasizes the importance of when throwing a timeout it is checked that the consensus runner is either part of the current epoch or was part of the prior epoch. 

This will be handled in a pr based on this branch: https://github.com/Concordium/concordium-node/tree/upon-timeout-event